SQL: XPack FeatureSet functionality

astefan opened the pull request:
This PR covers the requests in https://github.com/elastic/elasticsearch/issues/34821.
At the moment is still work in progress. Missing bits: `failed` queries metrics counters and REST requests differentiation between cli, native rest requests and other clients.

astefan committed:
Initial commit for XPack FeatureSet functionality

astefan committed:
Reverted wrong changes

elasticmachine commented:
Pinging @elastic/es-search-aggs

astefan committed:
Introduce "client" parameter for REST requests

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

astefan committed:
Rest test case

costin reviewed (changes requested):
The biggest issue with the current PR is that the hook is applied in a destructive fashion.
Instead of adding overloaded method types (which means different code paths), a dedicated rule should be used instead.

astefan commented on x-pack/plugin/sql/qa/multi-node/build.gradle:
```diff
@@ -5,6 +5,13 @@ description = 'Run a subset of SQL tests against multiple nodes'
  * feel should need to be tested against more than one node.
  */
+ 
+dependencies {
```
I have a new single node test that I will use instead. I'll add it shortly in a commit.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/stats/Counters.java:
```diff
@@ -0,0 +1,138 @@
+ * Calling toNestedMap() will create a nested map, where each dot of the key name will nest deeper
+ * The main reason for this class is that the stats producer should not be worried about how the map is actually nested
+ */
+public class Counters implements Streamable {
```
Changed it. Using the Watcher class now.

astefan committed:
Addressed review comments

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

astefan commented:
@costin thank you for your review. I've made changes and pushed a new version.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlStatsResponse.java:
```diff
@@ -0,0 +1,106 @@
+        return builder;
+    }
+
+    public static class Node extends BaseNodeResponse implements ToXContentObject {
```
I would rename it to something like `NodeSqlStats`

costin replied:
+1, we already have a `Node` inside the code base. `NodeStats`, `NodeSqlStats` etc...

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -1062,6 +1068,32 @@ private Expression implicitCast(Expression e) {
+                        metrics().inc(SUBSELECT);
+                    }
+                });
+                plan.forEachExpressionsDown(expr -> {
```
I think here we will count double if the subselect is also aliased or not?

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/stats/Metrics.java:
```diff
@@ -0,0 +1,108 @@
+    private final Map<FeatureMetric, CounterMetric> featuresMetrics = new HashMap<>();
+    // because the hook in the Verifier/Analyzer is being called several times during the query analysis flow
+    // we keep a set of flags so that we don't count a certain metric more than once
+    private final Map<FeatureMetric, Boolean> featuresCollected = new HashMap<>();
```
You could use a HashSet and check existence with `contains()`

costin replied:
Just use  `map.putIfAbsent()`

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -113,13 +116,13 @@ public Analyzer(FunctionRegistry functionRegistry, IndexResolution results, Time
+        return Arrays.asList(/*metricsCheck, */substitution, resolution);
     }
-
+    
```
Please revert.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -113,13 +116,13 @@ public Analyzer(FunctionRegistry functionRegistry, IndexResolution results, Time
         return analyze(plan, true);
     }
-
+    
```
Please revert.

matriv reviewed (commented):
Left some comments, thx!

costin commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/HttpClient.java:
```diff
@@ -92,7 +96,7 @@ public boolean queryClose(String cursor) throws SQLException {
         String query = "error_trace&mode=" +
                         request.mode() +
-                        (request.restClient() != null ? "&client=" + request.restClient().toString() : "");
+                        (request.clientId() != null ? "&clientid=" + request.clientId() : "");
```
Use a constant for client id (by declaring it inside `ConnectionConfiguration`).
To have consistent naming, use `client.id` instead of `clientid`

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java:
```diff
@@ -50,14 +50,16 @@ public PlanExecutor(Client client, IndexResolver indexResolver, NamedWriteableRe
         this.optimizer = new Optimizer();
         this.planner = new Planner();
+        
+        this.verifier = new Verifier();
```
The `Verifier` is used before the `Optimizer` so declare it as such including in the following method declarations.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -45,8 +49,18 @@ public RestSqlQueryAction(Settings settings, RestController controller) {
         try (XContentParser parser = request.contentOrSourceParamParser()) {
-            sqlRequest = SqlQueryRequest.fromXContent(parser, Mode.fromString(request.param("mode")),
-                                                      RestClient.fromString(request.param("client")));
+            String clientId = request.param("clientid");
```
Use constant for `client.id` instead of a string.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -45,8 +49,18 @@ public RestSqlQueryAction(Settings settings, RestController controller) {
+                if (clientId.length() > CLIENT_ID_MAX_LENGTH) {
+                    clientId = clientId.substring(0, CLIENT_ID_MAX_LENGTH).toLowerCase(Locale.ROOT);
+                }
+                if (!clientId.equals(CLI) && !clientId.equals(CANVAS)) {
```
Not sure what this block tries to achieve. If just `CLI` and `CANVAS` are accepted, why bother checking the length?

astefan committed:
Moved FeatureSetUsage to x-pack core and addressed some of the comments.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TransportSqlQueryAction.java:
```diff
@@ -12,37 +12,32 @@
     private final SqlLicenseChecker sqlLicenseChecker;
-    private static final Map<Metric, Map<String, CounterMetric>> opsByTypeMetrics = new ConcurrentHashMap<>();
-    private final Map<String, CounterMetric> featuresMetrics = new ConcurrentHashMap<>();
+    private static final Metrics metrics = new Metrics();
```
Why is the metric created in the transport and not inside `PlanExecutor`. Further more why is it static (what's the point of setting it as an instance).
There's only one `PlanExecutor` which is shared across all transports - doing the initialization in the transport means that if translate is called instead of query, the metric won't be inside `PlanExecutor`.
Further more transports are auxiliary to the executor and thus should just get the final Metric from the executor without any setting which implies that `PlanExecutor` can exist in an inconsistent state.

astefan replied:
The Metrics instance was static because it needed to be accessed in the static `operation()` method. Also, translate operations are not managed in Metrics, they are incremented in a separate transport action in a much simpler way.
I've moved the Metrics to the `PlanExecutor`.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TransportSqlQueryAction.java:
```diff
@@ -51,25 +46,7 @@ public TransportSqlQueryAction(TransportService transportService, ActionFilters
-        featuresMetrics.put("join",      new CounterMetric());
-        featuresMetrics.put("subselect", new CounterMetric());
-        this.planExecutor.setFeaturesMetrics(featuresMetrics);
+        planExecutor.metrics(metrics);
```
See above.

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/rule/RuleExecutor.java:
```diff
@@ -128,43 +125,17 @@ public TreeType after() {
-    
-    protected ExecutionInfo executeWithInfoWithMetrics(TreeType plan, Map<String, CounterMetric> featuresMetrics) {
-        return executeWithInfo(plan, featuresMetrics);
+        return executeWithInfo(plan).after;
```
Has the file been reverted to the original form? If that happens I would expect this file (and others) to not show up anymore.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/session/SqlSession.java:
```diff
@@ -40,27 +39,30 @@
     private final Configuration settings;
 
     public SqlSession(SqlSession other) {
         this(other.settings, other.client, other.functionRegistry, other.indexResolver,
```
Verifier should be between preAnalyzer and optimizer.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -1079,4 +1111,12 @@ protected boolean skipResolved() {
         }
     }
+    
+    abstract static class MetricsRule<SubPlan extends LogicalPlan> extends Rule<SubPlan, LogicalPlan> {
```
What value does this class add?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -93,7 +106,20 @@ private static Failure fail(Node<?> source, String message, Object... args) {
+        return failures.stream().collect(toMap(Failure::source, Failure::message));
+    }
+    
+    public void metrics(Metrics metrics) {
```
Move metrics into the `PlanExecutor` and pass it at construction time to the `Verifier`. There's no need for setters.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -42,10 +46,19 @@
+import static org.elasticsearch.xpack.sql.stats.FeatureMetric.WHERE;
+
+public final class Verifier {
+    private Metrics metrics;
```
`private final` - see below.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -212,6 +238,29 @@ private static Failure fail(Node<?> source, String message, Object... args) {
         }
+        
+        // gather metrics
+        if (failures.isEmpty() && metrics != null) {
```
To avoid incrementing the counter multiple times you have several options:
1. use `plan/expression.anyMatch()` for each check, this will result in multiple iterations though, one per check
2. use a method variable of a Map of booleans. This is essentially a `BitSet` where the ordinal of the enum turns on the feature is found.
After traversing the tree, traverse the `BitSet` and set the appropriate metric:
```java
BitSet b = new BitSet(Features.values().size());
...

if (p instanceof Aggregate) {
        b.set(GROUPBY.ordinal);
}
...

}
```

Depending on the carrotsearch or java.util BitSet you can have slightly different ways of iterating but essentially you find the index of the next set bit, convert that into an ordinal, get the enum and inc the metric.

```java
for( i = nextBit() ; i !=0; i = nextBit(i)) {
   metric.inc(Metrics.values()[i]);
}
```

astefan replied:
@costin thank you for your comments and valuable suggestions. I think both options can be used at the same time:
- `anyMatch()` to stop after the first match is found
- `BitSet` to replace the `Map<FeatureSet, Boolean>` current implementation in `Metrics`

But I cannot do the BitSet initialization in the Verifier, because the possible multiple increments happen because `Verifier.verify(LogicalPlan)` is called multiple times in the analysis process. So, the BitSet "control" should be outside the `Verifier`.

costin reviewed (changes requested):
The parsing can be simplified to avoid using maps through a simple, cheap bitset. Also the way the metric is passed around needs improvement.

astefan committed:
Moved Metrics inside the PlanExecutor and addressed review comments

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

astefan commented:
@costin @matriv pushed a new commit.

astefan committed:
Test fix

astefan committed:
Polishing
Bug fixed in the Analyzer
Unit testing for Metrics

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

astefan committed:
More polishing

astefan committed:
Cosmetic changes

astefan commented:
@matriv @costin I've pushed another change, hopefully complete now. I've removed the duplicate counting of metrics logic altogether and fixed the bug in the SqlSession that was triggering it (thanks @costin for the pointer), more polishing and cosmetic changes and also I've added unit testing for Metrics and more integration tests.

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/AbstractSqlRequest.java:
```diff
@@ -16,27 +16,35 @@
 public abstract class AbstractSqlRequest implements ToXContentFragment {
 
-    private final Mode mode;
+    private RequestInfo requestInfo;
```
`private final`

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/RequestInfo.java:
```diff
@@ -0,0 +1,55 @@
+        return mode;
+    }
+    
+    public void mode(Mode mode) {
```
Is this setter ever used? why not always use the constructor?

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/RequestInfo.java:
```diff
@@ -0,0 +1,55 @@
+    private String clientId;
+    
+    public RequestInfo(Mode mode) {
+        this.mode = mode;
```
`this(mode, null)`

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/RequestInfo.java:
```diff
@@ -0,0 +1,55 @@
+    public static String CLI = "cli";
+    public static String CANVAS = "canvas";
+    
+    private Mode mode;
```
both `final`

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/RequestInfo.java:
```diff
@@ -0,0 +1,55 @@
+        return clientId;
+    }
+    
+    public void clientId(String clientId) {
```
Same here, why setter?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -239,6 +238,9 @@ private static boolean containsAggregate(Expression exp) {
         return containsAggregate(singletonList(exp));
     }
 
+    private Metrics metrics() {
```
Why not eliminate it? If it has limited usage one can just call `verifier.metrics()`

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -212,6 +238,33 @@ private static Failure fail(Node<?> source, String message, Object... args) {
         }
+        
+        // gather metrics
+        if (failures.isEmpty() && metrics != null) {
```
Can metrics ever be null? Isn't this class always available inside PlanExecutor?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java:
```diff
@@ -34,8 +36,11 @@
     private final Optimizer optimizer;
     private final Planner planner;
+    
+    private final Metrics metrics = new Metrics();
```
To be consistent with the style, create the object inside the constructor.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -93,7 +110,16 @@ private static Failure fail(Node<?> source, String message, Object... args) {
+        return failures.stream().collect(toMap(Failure::source, Failure::message));
+    }
+    
+    public Metrics metrics() {
```
This needs to be removed.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlPlugin.java:
```diff
@@ -101,6 +104,15 @@ public SqlPlugin(Settings settings) {
+    @Override
+    public Collection<Module> createGuiceModules() {
+        List<Module> modules = new ArrayList<>();
+        modules.add(b -> {
```
`modules.add(b -> XPackPlugin.bindFeatureSet(b, SqlFeatureSet.class));`

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/session/SqlSession.java:
```diff
@@ -44,21 +46,23 @@
     public SqlSession(SqlSession other) {
         this(other.settings, other.client, other.functionRegistry, other.indexResolver,
-                other.preAnalyzer, other.optimizer,other.planner);
+                other.preAnalyzer, other.optimizer, other.planner, other.verifier);
```
Order should be: preAnalyzer/verifier/optimizer/planner

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/stats/VerifierMetricsTests.java:
```diff
@@ -0,0 +1,237 @@
+import static org.elasticsearch.xpack.sql.stats.FeatureMetric.WHERE;
+import static org.elasticsearch.xpack.sql.stats.Metrics.FPREFIX;
+
+public class VerifierMetricsTests extends ESTestCase {
```
👍

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/stats/Metrics.java:
```diff
@@ -0,0 +1,120 @@
+ */
+public class Metrics {
+    private enum OperationType {
+        FAILED, PAGING, TOTAL;
```
This needs to be broken into `QUERY`/`PAGING` each one with `FAILED`/`SUCCESSFUL/TOTAL`

costin reviewed (approved):
LGTM. Thanks for incorporating the feedback.
I've left some minor comments regarding style, nothing major.
There's the discussion of breaking the stats into queries/paging and each one total/failed/successful but if it's easier for you, that can be addressed in a follow-up PR.

matriv reviewed (approved):
LGTM. Great stuff! left some minor comments.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlStatsRequest.java:
```diff
@@ -0,0 +1,77 @@
+        return "sql_stats";
+    }
+    
+    static class Node extends BaseNodeRequest {
```
@matriv I was (still am) hesitant to rename this inner class because wherever it's used, it's prefixed with the enclosing class name: `return new SqlStatsRequest.Node(request, nodeId);` and it's already obvious that it belongs to the SqlSomething. Let me know if you are thinking about this differently.

astefan committed:
Addressed feedback

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

astefan commented:
run the gradle build tests 1

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 34821_feature
